packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,984 @@
+    }
+    return null;
+  }
+
+  Map<dynamic, dynamic> _decodeRestorationData(Uint8List data) {
+    if (data == null) {
+      return null;
+    }
+    final ByteData encoded = data.buffer.asByteData(data.offsetInBytes, data.lengthInBytes);
+    return const StandardMessageCodec().decodeMessage(encoded) as Map<dynamic, dynamic>;
+  }
+
+  Uint8List _encodeRestorationData(Map<dynamic, dynamic> data) {
+    final ByteData encoded = const StandardMessageCodec().encodeMessage(data);
+    return encoded.buffer.asUint8List(encoded.offsetInBytes, encoded.lengthInBytes);
+  }
+
+  bool _debugDoingUpdate = false;
+  bool _postFrameScheduled = false;
+
+  final Set<RestorationBucket> _bucketsNeedingSerialization = <RestorationBucket>{};
+
+  /// Called by a [RestorationBucket] to request serialization for that bucket.
+  ///
+  /// This method is called by a bucket in the hierarchy whenever the data
+  /// in it or the shape of the hierarchy has changed.
+  ///
+  /// Calling this is a no-op when the bucket is already scheduled for
+  /// serialization.
+  ///
+  /// It is exposed to allow testing of [RestorationBucket]s in isolation.
+  @protected
+  @visibleForTesting
+  void scheduleSerializationFor(RestorationBucket bucket) {
+    assert(bucket != null);
+    assert(bucket._manager == this);
+    assert(!_debugDoingUpdate);
+    _bucketsNeedingSerialization.add(bucket);
+    if (!_postFrameScheduled) {
+      _postFrameScheduled = true;
+      SchedulerBinding.instance.addPostFrameCallback((Duration _) => _doSerialization());
+    }
+  }
+
+  /// Called by a [RestorationBucket] to unschedule a request for serialization.
+  ///
+  /// This method is called by a bucket in the hierarchy whenever it no longer
+  /// needs to be serialized (e.g. because the bucket got disposed).
+  ///
+  /// It is safe to call this even when the bucket wasn't scheduled for
+  /// serialization before.
+  ///
+  /// It is exposed to allow testing of [RestorationBucket]s in isolation.
+  @protected
+  @visibleForTesting
+  void unscheduleSerializationFor(RestorationBucket bucket) {
+    assert(bucket != null);
+    assert(bucket._manager == this);
+    assert(!_debugDoingUpdate);
+    _bucketsNeedingSerialization.remove(bucket);
+  }
+
+  void _doSerialization() {
+    assert(() {
+      _debugDoingUpdate = true;
+      return true;
+    }());
+    _postFrameScheduled = false;
+
+    for (final RestorationBucket bucket in _bucketsNeedingSerialization) {
+      bucket.finalize();
+    }
+    _bucketsNeedingSerialization.clear();
+    sendToEngine(_encodeRestorationData(_rootBucket._rawData));
+
+    assert(() {
+      _debugDoingUpdate = false;
+      return true;
+    }());
+  }
+}
+
+/// Identifies a piece of data within or a child of a [RestorationBucket].
+///
+/// The [value] of a [RestorationId]s must be unique within a given
+/// [RestorationBucket]. [RestorationId] objects with the same [value] are
+/// considered equal.
+@immutable
+class RestorationId {
```

Comment:
Why not use a `String` directly?

Reply:
interesting question... i'm guessing the idea is that the class makes the API more self-documenting?

Reply:
After thinking about this, I've removed the RestorationId class and replaced it with a string. The class didn't add too much value...